src/IECoreImage/ImageReader.cpp

```diff
@@ -343,34 +326,44 @@ class ImageReader::Implementation
 		// Exception is thrown rather than false being returned.
 		bool open( bool throwOnFailure = false )
 		{
-			if( m_inputFile && m_reader->fileName() == m_inputFileName )
+			if( m_cache && m_reader->fileName() == m_inputFileName )
 			{
 				// we already opened the right file successfully
 				return true;
 			}
 
 			m_inputFileName = "";
-			m_inputFile.reset( ImageInput::open( m_reader->fileName() ) );
-			if( !m_inputFile )
+			m_cache.reset( ImageCache::create( /* shared */ false ) );
+
+			if( ImageInput *input = ImageInput::create( m_reader->fileName() ) )
 			{
-				if( !throwOnFailure )
-				{
-					return false;
-				}
-				else
-				{
-					throw IOException( string( "Failed to open file \"" ) + m_reader->fileName() + "\". " + geterror() );
-				}
+				m_cache->add_file( m_inputFileName );
+				m_inputFileName = m_reader->fileName();
+				m_fileFormat = input->format_name();
+				ImageInput::destroy( input );
+
+				return true;
 			}
 
-			m_inputFileName = m_reader->fileName();
+			if( !throwOnFailure )
+			{
+				return false;
+			}
+			else
+			{
+				throw IOException( string( "Failed to open file \"" ) + m_reader->fileName() + "\". " + geterror() );
+			}
+		}
 
-			return true;
+		static void destroyImageCache( ImageCache *cache )
+		{
+			ImageCache::destroy( cache, /* teardown */ true );
```

Comment:
The docs (and source) for OIIO seem to suggest that this teardown argument isn't relevant when the cache isn't shared anyway, so can we just use `ImageCache::destroy` directly rather than wrap it?

Reply:
When I tried to do it directly:

`std::unique_ptr<ImageCache, decltype(&ImageCache::destroy) > m_cache;`

I got this compile error:

```
src/IECoreImage/ImageReader.cpp:369:60: error: decltype cannot resolve address of overloaded function
   std::unique_ptr<ImageCache, decltype(&ImageCache::destroy) > m_cache;
```

The wrap seemed simple enough, unless you know the necessary c++11 shenanigans to get this to compile without?

Also, wrapping protects us in case the relevance of the teardown argument changes in the future.

Reply:
> The wrap seemed simple enough, unless you know the necessary c++11 shenanigans to get this to compile without?

I think `std::unique_ptr<ImageCache, void (*)( ImageCache * )>` should do it.

> Also, wrapping protects us in case the relevance of the teardown argument changes in the future.

I doubt that will ever happen. In the meantime, as someone trying to understand the code, I had to jump through another level of indirection, spot that the indirection was to pass a parameter I hadn't heard of, look up the documentation for that parameter, realise that it wasn't doing anything anyway and then try to figure out why this was necessary. It's definitely clearer just to use `ImageCache::destroy` directly...